cpp/src/arrow/util/concatenate.cc

```diff
@@ -0,0 +1,335 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include "arrow/util/concatenate.h"
+
+#include <algorithm>
+#include <limits>
+#include <memory>
+#include <utility>
+#include <vector>
+
+#include "arrow/array.h"
+#include "arrow/memory_pool.h"
+#include "arrow/util/logging.h"
+#include "arrow/util/visibility.h"
+#include "arrow/visitor_inline.h"
+
+namespace arrow {
+
+static inline ArrayData SliceData(const ArrayData& data, int64_t offset, int64_t length) {
+  DCHECK_LE(offset, data.length);
+  length = std::min(data.length - offset, length);
+  offset += data.offset;
+
+  auto copy = data;
+  copy.length = length;
+  copy.offset = offset;
+  copy.null_count = data.null_count != 0 ? kUnknownNullCount : 0;
+  return copy;
+}
+
+struct ConcatenateImpl {
+  ConcatenateImpl(const std::vector<ArrayData>& in, MemoryPool* pool)
+      : in_(in),
+        in_size_(static_cast<int>(in.size())),
+        lengths_(in.size()),
+        offsets_(in.size()),
+        pool_(pool) {
+    out_.type = in[0].type;
+    for (int i = 0; i != in_size_; ++i) {
+      lengths_[i] = in[i].length;
+      offsets_[i] = in[i].offset;
+      out_.length += lengths_[i];
```

Comment:
I'd invest in getting the proper shape of `out_` immediately, e.g. buffers{}, children_data{} ... This will remove a lot of the boiler plate in the Visit method of creating local buffers/array_data and pushing them. This can also be moved in the first Visit, _or_ create a new constructor (or static method) in ArrayData that does this book keeping.